libcasr/src/sarif.rs

```diff
@@ -0,0 +1,404 @@
+        let rule_id = if let Some(pos) = CLASSES
+            .iter()
+            .position(|item| item.1 == class.short_description)
+        {
+            format!("F{pos}")
+        } else {
+            let s = format!("G{}", self.id);
+            s
+        };
+
+        let rules = self.json.as_object().unwrap()["runs"].as_array().unwrap()[0]
+            .as_object()
+            .unwrap()["tool"]
+            .as_object()
+            .unwrap()["driver"]
+            .as_object()
+            .unwrap()["rules"]
+            .as_array()
+            .unwrap();
+        if rules
+            .iter()
+            .any(|r| r.as_object().unwrap()["name"].as_str().unwrap() == class.short_description)
+        {
+            // ExecutionClass is already added.
+            return (None, rule_id);
+        }
+        let mut rule = Map::new();
+        rule.insert("id".to_string(), Value::String(rule_id.clone()));
+        rule.insert(
+            "name".to_string(),
+            Value::String(class.short_description.clone()),
+        );
+        let mut short_desc = Map::new();
+        short_desc.insert("text".to_string(), Value::String(class.description.clone()));
+        rule.insert(
+            "shortDescription".to_string(),
+            Value::Object(short_desc.clone()),
+        );
+        let mut full_desc = Map::new();
+        full_desc.insert("text".to_string(), Value::String(class.explanation.clone()));
+        rule.insert(
+            "fullDescription".to_string(),
+            Value::Object(full_desc.clone()),
+        );
+        let mut properties = Map::new();
+        properties.insert(
+            "problem.severity".to_string(),
+            Value::String(class.severity.clone()),
+        );
+        rule.insert("properties".to_string(), Value::Object(properties));
+
+        (Some(Value::Object(rule)), rule_id)
+    }
+
+    /// Add CASR CrashReport to SARIF report.
+    /// NOTE: before using this method,
+    /// use SarifReprot::new() to get report.
+    ///
+    ///  # Arguments
+    ///
+    ///  * 'report' - CrashReport
+    ///
+    ///  * 'source_root' - Path to source root directory.
+    pub fn add_casr_report<T: AsRef<Path>>(
+        &mut self,
+        report: &CrashReport,
+        source_root: T,
+    ) -> Result<()> {
+        let (rule, rule_id) = self.rule(&report.execution_class);
+        if let Some(rule) = rule {
+            let rules = self.json.as_object_mut().unwrap()["runs"]
+                .as_array_mut()
+                .unwrap()[0]
+                .as_object_mut()
+                .unwrap()["tool"]
+                .as_object_mut()
+                .unwrap()["driver"]
+                .as_object_mut()
+                .unwrap()["rules"]
+                .as_array_mut()
+                .unwrap();
+            rules.push(rule);
+            if rule_id.starts_with('G') {
+                self.id += 1;
+            }
```

Comment:
we can move this code to `rule` function

Reply:
We couldn't be case it is &self.